### ISSUE TITLE
Covariant subtype return value "unknown this" incorrect conformance violation (NTI)

### ISSUE THREAD

**Issue by supersteves** [google/closure-compiler#1411]
I'm experimenting with this trick:
https://github.com/google/closure-compiler/wiki/covariant-subtype-values-with-@template

And I'm applying ```BanUnknownThis``` throughout my codebase:
https://github.com/google/closure-compiler/wiki/JS-Conformance-Framework
https://github.com/google/closure-compiler/blob/0f54fd1f1d51680d61fb40c647373349d536fcbb/src/com/google/javascript/jscomp/ConformanceRules.java#L1132

``` javascript
class X {
  /** 
   * @this {THIS}
   * @return {THIS}
   * @template THIS
   */
  x() {
    return this; // WARNING - Violation: unknown this
  }
}

class Y extends X {
}

var /** Y */ y = new Y().x();
```

NTI seems to correctly infer 'this' but the conformance violation warning still happens:

https://closure-compiler-debugger.appspot.com/#input0%3Dclass%2520X%2520%257B%250A%2520%2520%252F**%2520%250A%2520%2520%2520*%2520%2540this%2520%257BTHIS%257D%250A%2520%2520%2520*%2520%2540return%2520%257BTHIS%257D%250A%2520%2520%2520*%2520%2540template%2520THIS%250A%2520%2520%2520*%252F%250A%2520%2520x()%2520%257B%250A%2520%2520%2520%2520return%2520this%253B%2520%252F%252F%2520WARNING%2520-%2520Violation%253A%2520unknown%2520this%250A%2520%2520%257D%250A%257D%250A%250Aclass%2520Y%2520extends%2520X%2520%257B%250A%257D%250A%250Avar%2520%252F**%2520Y%2520*%252F%2520y%2520%253D%2520new%2520Y().x()%253B%26input1%26conformanceConfig%3Drequirement%253A%2520%257B%2520type%253A%2520CUSTOM%2520java_class%253A%2520'com.google.javascript.jscomp.ConformanceRules%2524BanUnknownThis'%2520error_message%253A%2520'unknown%2520this'%2520%257D%250A%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CHECK_TYPES_NEW_INFERENCE%3D1%26LANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1

Now is it the conformance rule that is broken or something else? Or am I doing it wrong? I don't see how NTI can be getting it right but ```BanUnknownThis``` still not working. Is ```BanUnknownThis``` just an example, or is it used in production?

**Comment by supersteves** [comment]
NB. I have such all warnings as fatal in my project, and suppress this erroneous warning using:
```@suppress {conformanceViolations}``` 

However this suppress tag requires a change to Closure Compiler; my tiny PR for this was rejected:
https://github.com/google/closure-compiler/pull/1199

In this case I am suppressing conformance violations on 3 specific methods, not on the entire (large in this case) source file. I know this is a bug, which may get fixed, but it might not for some time; isn't this an argument for accepting that one-line PR? I can't see how to exclude a few specific methods from conformance violations using the approach suggested.

**Comment by dimvar** [comment]
You are correct that NTI knows the type but the conformance check doesn't. The reason is that the compiler passes after NTI don't understand the new type system yet. So, after NTI runs, we also run the old type checker, and use the old types for the rest of the compilation. (And since the old type checker doesn't know the type in this case, the conformance check fails.)

**Comment by supersteves** [comment]
Got it. Presumably this can serve as the issue for updating the conformance check. Is this something others run into, and how do they suppress the warning (or do they permit warnings and ignore them)?

**Comment by SLaks** [comment]
I don't get why the conformance error happens here in the first place.
The conformance check explicitly ignores template types; see https://github.com/google/closure-compiler/blob/19acab07e9f1abdf0bd358a8646f05a07903f676/src/com/google/javascript/jscomp/ConformanceRules.java#L1231

Can that be fixed?

**Comment by dimvar** [comment]
It ignores template types in the property access, but not in the unknown-this check:
https://github.com/google/closure-compiler/blob/19acab07e9f1abdf0bd358a8646f05a07903f676/src/com/google/javascript/jscomp/ConformanceRules.java#L1198

The justification for ignoring them in the template access is because the unknown-this would catch the root violation:
https://github.com/google/closure-compiler/blob/19acab07e9f1abdf0bd358a8646f05a07903f676/src/com/google/javascript/jscomp/ConformanceRules.java#L1216
